### PR TITLE
Migrate kernel.js/paramz.js/unsafe.js to RacketScript

### PR DESCRIPTION
**Pull request by vishesh**
- Implement a minimal #lang racketscript/boot for writing kernel.js/...
-  For these primitive modules, avoid possible cycles automatically. This happens because certain macro introduced bindings would refer to original `#%kernel`
- Additions to interop and IL for handling `undefined`/`null` 

ISSUE #44

### CODE REVIEW

**Comment by vishesh** [comment]
@stchang All tests pass. This is a big patch, but if you get some time, please take a quick look. Mainly `racketscript-compiler/racketscript/boot.rkt` and `kernel.rkt`.

I've ported most JS code in kernel to RacketScript. What's left is Regex and few stuff you added recently. We just point to kernel.js implementation in that case.

Tests will run 2x slower for now :(

File: racketscript-compiler/racketscript/compiler/runtime/kernel.rkt [hunk 1]
```diff
@@ -0,0 +1,824 @@
...
+(define+provide (struct-type? v)
+  (#js.Core.Struct.isStructType v))
+
+;; --------------------------------------------------------------------------
+;; Vectors
+
+(define+provide vector
+  (v-λ ()
+     (#js.Core.Vector.make (#js.Core.argumentsToArray arguments) #t)))
+
```
Comment (stchang) [inline racketscript-compiler/racketscript/compiler/runtime/kernel.rkt]: add comment indicating optional arg

File: racketscript-compiler/racketscript/compiler/runtime/kernel.rkt [hunk 2]
```diff
@@ -0,0 +1,824 @@
...
+(define+provide (string-length v)
+  #js.v.length)
+
+(define+provide (string-downcase v)
+  (#js.v.toLowerCase v))
+
+(define+provide (string-upcase v)
+  (#js.v.toUpperCase v))
+
+(define+provide (substring str start end)
```
Comment (stchang) [inline racketscript-compiler/racketscript/compiler/runtime/kernel.rkt]: add comment for optional arg

**Comment by avelino** [comment]
PR merged, I think you can close this issue

**Pull request merged**